napari/_qt/dialogs/activity_dialog.py

```diff
@@ -0,0 +1,217 @@
+
+        Check whether pbar is nested and create ProgressBarGroup if it is, removing
+        existing separators and creating new ones. Show and start inProgressIndicator
+        to highlight to user the existence of a progress bar in the dock even when
+        the dock is hidden.
+
+        Parameters
+        ----------
+        pbar : ProgressBar
+            progress bar to add to activity dialog
+        nest_under : Optional[ProgressBar]
+            parent progress bar pbar should be nested under, by default None
+        """
+        if nest_under is None:
+            self._activityLayout.addWidget(pbar)
+        else:
+            # this is going to be nested, remove separators
+            # as the group will have its own
+            parent_pbar = nest_under._pbar
+            current_pbars = [parent_pbar, pbar]
+            remove_separators(current_pbars)
+
+            parent_widg = parent_pbar.parent()
+            # if we are already in a group, add pbar to existing group
+            if isinstance(parent_widg, ProgressBarGroup):
+                nested_layout = parent_widg.layout()
+            # create ProgressBarGroup for this pbar
+            else:
+                new_group = ProgressBarGroup(nest_under._pbar)
+                new_group.destroyed.connect(self.maybe_hide_progress_indicator)
+                nested_layout = new_group.layout()
+                self._activityLayout.addWidget(new_group)
+            new_pbar_index = nested_layout.count() - 1
+            nested_layout.insertWidget(new_pbar_index, pbar)
+
+        # show progress indicator and start gif
+        self._toggleButton._inProgressIndicator.movie().start()
+        self._toggleButton._inProgressIndicator.show()
+        pbar.destroyed.connect(self.maybe_hide_progress_indicator)
+
+    def move_to_bottom_right(self, offset=(8, 8)):
+        """Position widget at the bottom right edge of the parent."""
+        if not self.parent():
+            return
+        sz = self.parent().size() - self.size() - QSize(*offset)
+        self.move(QPoint(sz.width(), sz.height()))
+
+    def maybe_hide_progress_indicator(self):
+        """Hide progress indicator only if no progress bars are visible."""
+        pbars = self._baseWidget.findChildren(ProgressBar)
+        pbar_groups = self._baseWidget.findChildren(ProgressBarGroup)
+
+        progress_visible = any([pbar.isVisible() for pbar in pbars])
+        progress_group_visible = any(
+            [pbar_group.isVisible() for pbar_group in pbar_groups]
+        )
+        if not progress_visible and not progress_group_visible:
+            self._toggleButton._inProgressIndicator.movie().stop()
+            self._toggleButton._inProgressIndicator.hide()
+
+
+def get_pbar(nest_under=None, **kwargs):
+    """Adds ProgressBar to viewer ActivityDialog and returns it.
+    If nest_under is valid ProgressBar, nests new bar underneath
```

Comment:
```suggestion
    """Adds ProgressBar to viewer ActivityDialog and returns it.

    If nest_under is valid ProgressBar, nests new bar underneath
```

Reply:
I think this function should take in a viewer as an optional parameter, and only resort to QtMainWindow if it is not given.